Add compileTime macro to measure compilation times

fthomas opened the pull request:
`compileTime` is a simple macro that measures the compilation
time of a code fragment. Like `illTyped` it takes Scala code as
string, measures the time it takes to parse and typecheck it and
returns that time as a `Duration`.

Example usage:
```scala
scala> test.compileTime("Generic[(Int, Option[String])]")
res0: scala.concurrent.duration.FiniteDuration = 34911863 nanoseconds
```

milessabin commented:
Awesome! Thank you!

Seeing as there isn't (and can't easily be) a test for this, would you mind putting your description above into a doc comment?

InTheNow commented:
For the sake of [scoverage that will land soon ](https://github.com/milessabin/shapeless/pull/551), is it perhaps worth adding a "test" that just calls the "code" (including bad code) and checks for a duration `> 0`?

fthomas commented:
Yes, that is a good idea. I'll add a test and improve the docstring tonight.

fthomas commented:
I've added a test and updated the documentation. Unfortunately the build for 2.10 failed because it timed out. However this looks unrelated to this change since the build was aborted during `packageLocal`. Could you please rebuild or increase the timeout?

fthomas commented:
I increased the timeout from 40 to 45 minutes and now it has successfully built in under 30 minutes. :-)

milessabin commented:
Awesome! Thanks. Yes, the Travis timeout has been borderline for a little while now.

fommil commented:
@fthomas this could be really amazing in ENSIME to disable slow files. Can you please take a look at https://github.com/ensime/ensime-server/issues/954 and see if you can come up with any ideas?

fommil commented:
btw, I'm aware this is for code snippets. I was thinking in terms of writing a compiler plugin along the lines of what Jason recommends in that ticket for the source code files that are being compiled.

fthomas commented:
@fommil I looked at that ticket but that is way above my head.
